src/datachain/lib/dc.py

```diff
@@ -1624,6 +1625,208 @@ def subtract(  # type: ignore[override]
+                deleted=True,
+                modified=True,
+                unchanged=True,
+                status_col="diff"
+            )
+            ```
+        """
+        rname = "right_"
+
+        def _rprefix(c: str, rc: str) -> str:
+            """Returns prefix of right of two companion left - right columns
+            from merge. If companion columns have the same name then prefix will
+            be present in right column name, otherwise it won't.
+            """
+            return rname if c == rc else ""
+
+        def _to_list(obj: Union[str, Sequence[str]]) -> list[str]:
+            return [obj] if isinstance(obj, str) else list(obj)
+
+        if on is None:
+            raise ValueError("'on' must be specified")
+
+        on = _to_list(on)
+        if right_on:
+            right_on = _to_list(right_on)
+            if len(on) != len(right_on):
+                raise ValueError("'on' and 'right_on' must be have the same length")
+
+        if compare:
+            compare = _to_list(compare)
+
+        if right_compare:
+            if not compare:
+                raise ValueError(
+                    "'compare' must be defined if 'right_compare' is defined"
+                )
+
+            right_compare = _to_list(right_compare)
+            if len(compare) != len(right_compare):
+                raise ValueError(
+                    "'compare' and 'right_compare' must be have the same length"
+                )
+
+        num_statuses = sum(1 if s else 0 for s in [added, deleted, modified, unchanged])
+
+        if num_statuses > 1 and not status_col:
+            raise ValueError(
+                "Status column name is needed if more than one status is asked"
+            )
+
+        if num_statuses == 0:
+            raise ValueError(
+                "At least one of added, deleted, modified, unchanged flags must be set"
+            )
+
+        # we still need status column for internal implementation even if not
+        # needed in output
+        status_col = status_col or "sys__diff"
+
+        # calculate on and compare column names
+        right_on = right_on or on
+        cols = self.signals_schema.db_signals()
+        right_cols = other.signals_schema.db_signals()
+
+        on = self.signals_schema.resolve(*on).db_signals()  # type: ignore[assignment]
+        right_on = other.signals_schema.resolve(*right_on).db_signals()  # type: ignore[assignment]
+        if compare:
+            right_compare = right_compare or compare
+            compare = self.signals_schema.resolve(*compare).db_signals()  # type: ignore[assignment]
+            right_compare = other.signals_schema.resolve(*right_compare).db_signals()  # type: ignore[assignment]
+        else:
+            compare = [c for c in cols if c in right_cols]  # type: ignore[misc, assignment]
+            right_compare = compare
+
+        diff_cond = []
+
+        if added:
+            added_cond = sqlalchemy.and_(
```

Comment:
Can we replace all sqlalchamy function to our built-in functions.
It can look like this:

```python
from datachain.func import _case, _if

l = dc1.mutate(ldiff=1)
r = dc2.mutate(rdiff=1)

dc_diff = (
 l.join(r, on="id")
 .mutate(d1=_case(_isnon(ldiff), "A", "D") )
 .mutate(d2=_case(_isnon(rdiff), "A", "D") )
 .mutate(diff=_case(d1!=d2, "A", "D") )
 .select_except("d1", "d2", "ldiff", "rdiff")
)
```

Reply:
As we talked, I will create a followup issue for this as improvement, not to block additional work on delta updated etc.